src/Compilers/CSharp/Portable/Symbols/Extensions/SourceExtensionImplementationMethodSymbol.cs

```diff
@@ -0,0 +1,154 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Emit;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    internal sealed class SourceExtensionImplementationMethodSymbol : RewrittenMethodSymbol
+    {
+        public const string StaticExtensionNamePrefix = "<StaticExtension>";
+        public const string InstanceExtensionNamePrefix = "<Extension>";
+
+        public SourceExtensionImplementationMethodSymbol(MethodSymbol sourceMethod) : base(sourceMethod, TypeMap.Empty)
+        {
+            Debug.Assert(sourceMethod.ContainingSymbol is NamedTypeSymbol { IsExtension: true });
+            Debug.Assert(!sourceMethod.IsExtern);
+            Debug.Assert(!sourceMethod.IsExternal);
+        }
+
+        protected override void CreateTypeParameters(MethodSymbol originalMethod, ref TypeMap typeMap, out ImmutableArray<TypeParameterSymbol> typeParameters)
+        {
+            // PROTOTYPE: Are we creating type parameters with the right emit behavior? Attributes, etc.
+            //            Also, they should be IsImplicitlyDeclared
+            typeMap = typeMap.WithAlphaRename(originalMethod.ContainingType.TypeParameters.Concat(originalMethod.TypeParameters), this, out typeParameters);
+        }
+
+        public override int Arity => TypeParameters.Length;
+
+        public override string Name
+        {
+            get
+            {
+                return (_originalMethod.IsStatic ? StaticExtensionNamePrefix : InstanceExtensionNamePrefix) + _originalMethod.Name;
+            }
+        }
+
+        public override MethodKind MethodKind => MethodKind.Ordinary;
+        public override bool IsImplicitlyDeclared => true;
+
+        internal override bool HasSpecialName => true;
+
+        internal override int ParameterCount
+        {
+            get
+            {
+                return _originalMethod.ParameterCount +
+                    (_originalMethod.IsStatic || _originalMethod.ContainingType is not SourceNamedTypeSymbol { ExtensionParameter: { } } ? 0 : 1);
```

Comment:
Would it be better to not create an implementation method in the first place when the `ExtensionParameter` is faulty?

Reply:
I do not know. So far that didn't cause an problems.